File: HW9/task2.py
import random
import sys

SEPARATOR = "\n"
UNICODE = "utf-8"


class Node(object):
    def __init__(self, key, prior):
        self.key = key
        self.prior = prior
        self.parent = None
        self.height = 0
        self.c = 0
        self.l = None
        self.r = None


class Treap(object):
    nodes = set()

    @staticmethod
    def gen():
        return random.randint(1, 1000)

    def __init__(self):
        self.root = None

    def split(self, t, key):
        if t is None:
            return (None, None)
        if key <= self.get_height(t.l):
            l, t.l = self.split(t.l, key)
            r = t
            r.c = self.get_height(r)
            return (l, r)
        else:
            t.r, r = self.split(t.r, key - self.get_height(t.l) - 1)
            l = t
            l.c = self.get_height(l)
            return (l, r)

        return (l, r)


    def merge(self, l, r):
        if l is None:
            return r
        elif r is None:
            return l
        elif l.prior > r.prior:
            l.r = self.merge(l.r, r)
            l.c = self.get_height(l)
            return l
        else:
            r.l = self.merge(l, r.l)
            r.c = self.get_height(r)
            return r

    def find(self, t, key):
        if t is None:
            return "false"
        if t.key == key:
            return "true"
        elif key < t.key:
            return self.find(t.l, key)
        else:
            return self.find(t.r, key)

    def next(self, key):
        res = "none"
        t = self.root
        while t:
            if t.key <= key:
                t = t.r
            elif t.key > key:
                res = t.key
                t = t.l
        return str(res)

    def prev(self, key):
        res = "none"
        t = self.root
        while t:
            if t.key < key:
                res = t.key
                t = t.r
            elif t.key >= key:
                t = t.l
        return str(res)

    def insert(self, key):
        if key in self.nodes:
            return
        self.nodes.add(key)
        new_node = Node(key, self.gen())
        if self.root is None:
            self.root = new_node
            return
        l, r = self.split(self.root, key - 1)
        self.root = self.merge(l, new_node)
        self.root = self.merge(self.root, r)

    def remove(self, key):
        # if key not in self.nodes:
        #     return
        # self.nodes.remove(key)
        # l, r = self.split(self.root, key)
        # l1, _ = self.split(l, key - 1)
        # self.root = self.merge(l1, r)
        l, r = self.split(self.root, key)
        _, r1 = self.split(r, 1)
        self.root = self.merge(l, r1)

    def add(self, pos, key):
        l, r = self.split(self.root, pos)
        self.root = self.merge(self.merge(l, Node(key, self.gen())), r)


    def to_front(self, lidx, ridx):
        l, r = self.split(self.root, ridx + 1)
        l1, r1 = self.split(l, lidx)
        self.root = self.merge(self.merge(r1, l1), r)

    def get_height(self, t):
        if t is None:
            return 0
        # ans = 1
        # ans = max(ans, self.get_height(t.l) + 1)
        # ans = max(ans, self.get_height(t.r) + 1)
        # t.height = ans
        # ans =
        return 1 + self.get_height(t.l) + self.get_height(t.r)

    def pr(self, t, lst):
        if t is None:
            return
        if t.l:
            self.pr(t.l, lst)
        # print(t.key, end=" ")
        lst.append(t.key)
        if t.r:
            self.pr(t.r, lst)

if __name__ == "__main__":
    treap = Treap()
    n, m = map(int, sys.stdin.readline().split())
    lst = [int(i) for i in sys.stdin.readline().split()]
    for i in range(len(lst)):
        treap.add(i, lst[i])
    while m:
        line = sys.stdin.readline().split()
        if line[0] == "del":
            key = int(line[1]) - 1
            treap.remove(key)
        elif line[0] == "add":
            pos, key = map(int, line[1:])
            treap.add(pos, key)
        m -= 1
    lst = []
    treap.pr(treap.root, lst)
    sys.stdout.write(str(len(lst)) + "\n")
    sys.stdout.write(" ".join([str(i) for i in lst]))
    # n = 100
    # commands = ["insert", "delete"]
    # values = list(range(25))
    # while n:
    #     # try:
    #     #     command, value = input().split()
    #     # except:
    #     #     break
    #     command = random.heighthoice(commands)
    #     value = random.heighthoice(values)
    #     print(command, value)
    #
    #     if command == "insert":
    #         treap.insert(int(value))
    #         treap.get_height(treap.root)
    #         treap.pr(treap.root)
    #         print()
    #     elif command == "delete":
    #         treap.remove(int(value))
    #         treap.get_height(treap.root)
    #         treap.pr(treap.root)
    #         print()
    #     # elif command == "exists":
    #     #     print(treap.find(treap.root, int(value)))
    #     # elif command == "next":
    #     #     print(treap.next(int(value)))
    #     # elif command == "prev":
    #     #     print(treap.prev(int(value)))
    #     n -= 1

#
# if __name__ == "__main__":
#     treap = Treap()
#     data = sys.stdin.buffer.readlines()
#     result_exists = []
#     n_operations = len(data)
#     for i in range(n_operations):
#         command, value = data[i].decode(UNICODE).split()
#         if command == "insert":
#             treap.insert(int(value))
#         elif command == "delete":
#             treap.remove(int(value))
#         elif command == "exists":
#             result_exists.append(treap.find(treap.root, int(value)))
#         elif command == "next":
#             result_exists.append(treap.next(int(value)))
#         elif command == "prev":
#             result_exists.append(treap.prev(int(value)))
#     encoded_array = (SEPARATOR.join(result_exists)).encode(UNICODE)
#     sys.stdout.buffer.write(encoded_array)

File: HW9/task2.cpp
#include <iostream>
#include <vector>
#include <string>


struct Node {
	int y;
	int sz;
	int val;
	bool reverse;
	Node *l, *r;
};

Node *new_node(int val) {
	Node *result = new Node;
	result->y = std::rand();
	result->sz = 1;
	result->val = val;
	result->l = result->r = nullptr;
	result->reverse = false;
	return result;
}

class Treap {
private:
	Node *root;

	int get_sz(Node *t) {
		if (t == nullptr) {
			return 0;
		}
		return t->sz;
	}

	void upd_sz(Node *t) {
		if (t == nullptr) {
			return;
		}
		t->sz = 1 + this->get_sz(t->l) + this->get_sz(t->r);
	}

	void push(Node *it) {
		if (it && it->reverse) {
			it->reverse = false;
			std::swap(it->l, it->r);
			if (it->l) {
			    it->l->reverse ^= true;
			}
			if (it->r) {
			    it->r->reverse ^= true;}
		}
	}

	Node *merge(Node *t1, Node *t2) {

		if (t1 == nullptr) {
		    return t2;
		 }
		if (t2 == nullptr) {
		    return t1;
		    }
		this->push(t1);
		this->push(t2);
		if (t1->y > t2->y) {
			t1->r = this->merge(t1->r, t2);
			this->upd_sz(t1);
			return t1;
		} else {
			t2->l = this->merge(t1, t2->l);
			this->upd_sz(t2);
			return t2;
		}
	}

	void split(Node *t, int x, Node *&t1, Node *&t2) {
		if (t == nullptr) {
			t1 = t2 = nullptr;
			return;
		}
		this->push(t);
		if (get_sz(t->l) < x) {
			this->split(t->r, x - this->get_sz(t->l) - 1, t->r, t2);
			t1 = t;
		} else {
			this->split(t->l, x, t1, t->l);
			t2 = t;
		}
		this->upd_sz(t);
	}

	Node *add(Node *t, int pos, int val) {
		Node *t1, *t2;
		this->split(t, pos, t1, t2);
		Node *new_tree = new_node(val);
		t1 = this->merge(t1, new_tree);
		return this->merge(t1, t2);
	}

	Node *remove(Node *t, int pos) {
		Node *t1, *t2, *t3, *t4;
		this->split(t, pos, t1, t2);
		this->split(t2, 1, t3, t4);
		t = this->merge(t1, t4);
		delete t3;
		return t;
	}

	int get_value(Node *t, int pos) {
		int my_idx = this->get_sz(t->l);
		if (pos < my_idx) {
			return this->get_value(t->l, pos);
		} else if (pos == my_idx) {
			return t->val;
		} else {
			return this->get_value(t->r, pos - my_idx - 1);
		}
	}

	Node *to_front(Node *t, int l, int r) {
		Node *t1, *t2, *t3, *t4;
		this->split(t, r + 1, t1, t2);
		this->split(t1, l, t3, t4);
		return this->merge(this->merge(t4, t3), t2);
	}

	void print_tree(Node *t) {
		if (t == nullptr) {
		    return;
		}
		this->push(t);
		this->print_tree(t->l);
		std::cout << t->val << " ";
		this->print_tree(t->r);
	}

	Node *reverse(Node *t, int l, int r) {
		Node *t1, *t2, *t3;
		this->split(t, l, t1, t2);
		this->split(t2, r - l + 1, t2, t3);
		t2->reverse ^= true;
		return this->merge(this->merge(t1, t2), t3);
	}

public:
	Treap() { root = nullptr; }

	void from_vector(const std::vector<int> &v) {
		Node *result = nullptr;
		for (int i : v) {
			result = this->merge(result, new_node(i));
		}
		this->root = result;
	}

	void add(int pos, int val) {
		this->root = this->add(this->root, pos, val);
	}

	void remove(int pos) {
		this->root = this->remove(this->root, pos);
	}

	void print_tree() {
		this->print_tree(this->root);
	}

	int get_sz() {
		return this->get_sz(this->root);
	}

	void to_front(int l, int r) {
		this->root = this->to_front(this->root, l, r);
	}

	void reverse(int l, int r) {
		this->root = this->reverse(this->root, l, r);
	}
};

int main() {
	int n, m;
	std::cin >> n >> m;
	std::vector<int> v_arr(n);
	for (int i = 0; i < n; i++) {
		std::cin >> v_arr[i];
	}
	Treap treap = Treap();
	treap.from_vector(v_arr);
	std::string command;
	int pos, value;
	while (m--) {
		std::cin >> command;
		if (command == "add") {
			std::cin >> pos >> value;
			treap.add(pos, value);
		} else if (command == "del") {
			std::cin >> value;
			treap.remove(value - 1);
		}
	}
	std::cout << treap.get_sz() << "\n";
	treap.print_tree();
	return 0;
}